### PR TITLE
[JENKINS-59180] Allow AMI filtering instead of AMI ID

### PR DESCRIPTION
**Pull request by dbnicholson**
Allow finding the image to launch by the various `describeImages` criteria instead of just a specific image ID. Currently you're required to reconfigure the plugin whenever the image is updated, which is inconvenient.

Fixes: https://issues.jenkins.io/browse/JENKINS-59180

### CODE REVIEW

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (froblesmartin) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:1269]: Sorry if I am wrong but, Is this sentence right? 
AFAI understand, the request could have no AMI but only filters, right? So it would be something like `"Getting images with request: " + request`?

**Review by froblesmartin — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (dbnicholson) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java:1269]: Yeah, that's better.

**Review by froblesmartin — APPROVED** [review]
LGTM but I am still pretty new. Would like a review from @res0nance @fcojfernandez or @MRamonLeon

File: src/main/java/hudson/plugins/ec2/EC2Filter.java
Comment (fcojfernandez) [inline src/main/java/hudson/plugins/ec2/EC2Filter.java:74]: should be `values = "value1,value2,value3"` equals to `values="value2,value1,value3"`? And what about `values="value1, value2, value3"? According to `toFilter` method, these `equals` and `hashCode` are going to work and you might get duplicated Filter objects.

From UI I see you're allowing a list of EC2Filter

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (fcojfernandez) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:840]: why this change? I don't see the relationship with the desired fix and I'd appreciate if you could elaborate. Thanks!

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (fcojfernandez) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:908]: I don't think this concrete change should be an issue since I don't think this package is used out of this plugin, but it's a good practice in the community to deprecate the method and create the new method with the signature changes. The deprecated method can be removed in some versions.
This good practice is applied to all non-private and non-restricted methods just in case there is any proprietary plugin from any user that is extending this code. -> FYI @froblesmartin

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (fcojfernandez) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:911]: ditto

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (fcojfernandez) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:1242]: Sorry if my question is a nonsense, but any reason to use `Util.fixEmpty` instead of `fixEmptyAndTrim`? if ami is "   ", that would result on a List with the unique value "   "
```suggestion
        List<String> imageIds = fixEmptyAndTrim(ami) == null ?
```

**Review by fcojfernandez — CHANGES_REQUESTED** [review]
See my comment about changing a signature of a non-private method

Besides, if this PR intends to fix https://issues.jenkins.io/browse/JENKINS-59180, shouldn't be the ticket assigned to you and in progress?

File: src/main/java/hudson/plugins/ec2/EC2Filter.java
Comment (dbnicholson) [reply src/main/java/hudson/plugins/ec2/EC2Filter.java:74]: Good call, I didn't think of that. It would be better if equality tested the split tokens instead of the raw string.

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (dbnicholson) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java:840]: Because the image ID can be dynamic are this PR. In order to list the image ID, you'd need to do a round trip to AWS every time this is called. Which is actually often because `logProvisionInfo` uses it.

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (dbnicholson) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java:908]: Good call. I hadn't thought of any of the methods besides the constructor being used outside of the package.

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (dbnicholson) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java:1242]: I think because the plugin was already taking whatever was in `ami` I had decided that the only thing I wanted to detect was the empty string that would result from the form being filled in. But I agree that trimming would be more robust.

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (dbnicholson) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java:911]: 👍

**Comment by dbnicholson** [comment]
> Besides, if this PR intends to fix https://issues.jenkins.io/browse/JENKINS-59180, shouldn't be the ticket assigned to you and in progress?

I didn't think I could do anything more than make comments on there as an outside contributor. I've changed that now.

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (dbnicholson) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java:840]: To expand on that a bit more, you could just leave it as is but in the case where you were using filters it would just say `ami=""`, which wouldn't be helpful if you were looking in the logs. I actually think this is an improvement generally since you could have multiple templates sharing the same image ID. This way you can differentiate them if you use a unique description for each template.

**Review by fcojfernandez — APPROVED** [review]
Thank you very much for your contribution @dbnicholson
> I didn't think I could do anything more than make comments on there as an outside contributor. I've changed that now.

Please, feel free and assign any open ticket you decide to work on. That way you will let others
know you are on it.

And let me another piece of advice: try to avoid the force-push command as much as you can. It makes harder review the new changes since the force-push hides them. There are even reviewers that stop reviewing the PRs after a force-push :( 
 
@froblesmartin my concerns are addressed, so I'm approving the PR. I'll let you merge it just in case you want to have a second look at the latest changes.

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java:1285]: We should document this behaviour the the help text

**Review by res0nance — APPROVED** [review]
Untested with an aws account but overall looks good

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java
Comment (dbnicholson) [reply src/main/java/hudson/plugins/ec2/SlaveTemplate.java:1285]: It does say that in the form help for each field, but maybe you mean somewhere else?

**Comment by dbnicholson** [comment]
> And let me another piece of advice: try to avoid the force-push command as much as you can. It makes harder review the new changes since the force-push hides them. There are even reviewers that stop reviewing the PRs after a force-push :(

It's unfortunately the only way to properly maintain commit history on GitHub. That or someone applies the fixups before merging. Do you have something that does that? Or maybe you do squashed merges? I'm fine following whatever flow you normally use. There is a link on the text that says force push that shows the differences from the previous state.

**Comment by froblesmartin** [comment]
@dbnicholson is there anything else you want to change before merging, or may I merge already? 😉

**Comment by dbnicholson** [comment]
Nothing more from me. Go ahead and merge :smiley_cat:

**Comment by aviadcye** [comment]
Hi,
Sorry, but no matter what I do, I can't seem to get this new configuration to work. I always get the error "The request must contain the parameter ImageId". However if I set the ImageId, all other filters are ignored... Am I missing something?

Steps to reproduce.

1. Retrofitting an existing AMI configuration that uses a specific ImageId.
2. Removed the "AMI ID" set, so the filed is blank.
3. Set "AMI Owners" & "AMI Users" to self (as per example test in code).
4. Set "AMI Filters", Added just "Name" (also tried all lower case) with the value of the "Jenkins_docker_agent_base_-_7.12.20a". Also tired with "Jenkins_docker_agent_base*".
5. Click on "Check AMI".

Thank you, Aviad.

**Comment by dbnicholson** [comment]
Does your AWS account have an AMI named `Jenkins_docker_agent_base_-_7.12.20a`? Using an owner of `self` says your account is the owner of the AMI.

**Comment by dbnicholson** [comment]
Also, if you look in the logs, you should see the results of the `DescribeImages` call.

**Comment by aviadcye** [comment]
> Does your AWS account have an AMI named `Jenkins_docker_agent_base_-_7.12.20a`? Using an owner of `self` says your account is the owner of the AMI.

Yes i have this AMI in my account.



> Also, if you look in the logs, you should see the results of the DescribeImages call.

So when I use "Name" i get the error "The filter 'Name' is invalid" (the tag key is "Name", so I don't get why... but OK...
So i tried using "name", and I just didn't get a log entry.

**Comment by dbnicholson** [comment]
`name` and a tag named `Name` are different things. If you want to filter on a `Name` tag, you have to use a filter of `tag:Name`. This is in the AWS documentation linked to in the help text.

**Comment by aviadcye** [comment]
I tried that as well... didn't  work... :/
Btw, I think an example in the plugin context documentation before linking
to the aws docs would help...

**Comment by dbnicholson** [comment]
Can you post the logs with the `DescribeImages` request and response? Have you tried looking up this combination with some like awscli?

**Comment by aviadcye** [comment]
As I said, when I use the "check ami" button, there are no logs unless it
has a filter error, and I'm not even sure ab that.
I'll try to create the definition, and simply attempt to create the
agent... maybe that will give logs.
About awscli, ill try that as soon as possible and report back.

**Comment by dbnicholson** [comment]
The Check AMI button only works if you supply an image ID due to limitations in Jenkins. I thought that was clear that the button was not relevant when using filters.

**Comment by aviadcye** [comment]
Well this is interesting...
Even though the "Check AMI" button, still doesn't work as expected, trying to spawn an agent, worked on the first go...
That is, with simply leaving the imageid blank and adding the filter of "tag:Name = Jenkins_docker_agent_base", I was able to create an agent (logs below).

So this is just a lacking in the check AMI button functionality. 
IMHO, it should both be able to do the same as spawning an agent can, and provide logs... because I gather now that the log I got about the filter "Name" no being valid, was after I had saved the configuration.

At any rate, this is totally usable as it is right now.
Thank you for an awesome feature, I'll start using it immediately :)

`2020-12-31 15:01:36.746+0000 [id=140760]	INFO	hudson.plugins.ec2.SlaveTemplate#getImage: Getting image for request {ExecutableUsers: [],Filters: [{Name: tag:Name,Values: [Jenkins_docker_agent_base*]}],ImageIds: [],Owners: []}
2020-12-31 15:01:37.557+0000 [id=140760]	INFO	hudson.plugins.ec2.SlaveTemplate#logProvisionInfo: SlaveTemplate{description='test', labels=''}. Considering launching
2020-12-31 15:01:37.557+0000 [id=140760]	INFO	hudson.plugins.ec2.SlaveTemplate#logProvisionInfo: SlaveTemplate{description='test', labels=''}. Setting Instance Initiated Shutdown Behavior : ShutdownBehavior.Terminate
2020-12-31 15:01:37.728+0000 [id=140760]	INFO	hudson.plugins.ec2.SlaveTemplate#logProvisionInfo: SlaveTemplate{description='test', labels=''}. Looking for existing instances with describe-instance: {Filters: [{Name: image-id,Values: [ami-yepyepeyep]}, {Name: instance-type,Values: [t1.micro]}, {Name: key-name,Values: [jenkins-slaves]}, {Name: tag:jenkins_server_url,Values: [https://jenkins.dev.example.com/]}, {Name: tag:jenkins_slave_type,Values: [demand_test]}],InstanceIds: [],}
2020-12-31 15:01:39.154+0000 [id=140760]	INFO	hudson.plugins.ec2.SlaveTemplate#logProvisionInfo: SlaveTemplate{description='test', labels=''}. Return instance: {AmiLaunchIndex: 0,ImageId: ami-yepyepeyep,InstanceId: i-idididid,InstanceType: t1.micro,KeyName: jenkins-slaves,LaunchTime: Thu Dec 31 15:01:38 GMT 2020,Monitoring: {State: disabled},Placement: {AvailabilityZone: eu-west-1a,GroupName: ,Tenancy: default,},PrivateDnsName: ip-172-31-19-43.eu-west-1.compute.internal,PrivateIpAddress: 172.31.19.43,ProductCodes: [],PublicDnsName: ,State: {Code: 0,Name: pending},StateTransitionReason: ,SubnetId: subnet-09271c40,VpcId: vpc-031c1064,Architecture: x86_64,BlockDeviceMappings: [],ClientToken: 59eb1abe-d117-4183-bfff-b674b60fb26f,EbsOptimized: false,EnaSupport: true,Hypervisor: xen,ElasticGpuAssociations: [],ElasticInferenceAcceleratorAssociations: [],NetworkInterfaces: [{Attachment: {AttachTime: Thu Dec 31 15:01:38 GMT 2020,AttachmentId: eni-attach-045b1c07fe3064f89,DeleteOnTermination: true,DeviceIndex: 0,Status: attaching},Description: ,Groups: [{GroupName: default,GroupId: sg-c86505b3}],Ipv6Addresses: [],MacAddress: 06:54:38:0d:a6:0f,NetworkInterfaceId: eni-0e558a4d70a8bd16f,OwnerId: actactact,PrivateDnsName: ip-172-31-19-43.eu-west-1.compute.internal,PrivateIpAddress: 172.31.19.43,PrivateIpAddresses: [{Primary: true,PrivateDnsName: ip-172-31-19-43.eu-west-1.compute.internal,PrivateIpAddress: 172.31.19.43}],SourceDestCheck: true,Status: in-use,SubnetId: subnet-09271c40,VpcId: vpc-031c1064,InterfaceType: interface}],RootDeviceName: /dev/sda1,RootDeviceType: ebs,SecurityGroups: [{GroupName: default,GroupId: sg-c86505b3}],SourceDestCheck: true,StateReason: {Code: pending,Message: pending},Tags: [{Key: jenkins_slave_type,Value: demand_test}, {Key: jenkins_server_url,Value: https://jenkins.dev.example.com/}],VirtualizationType: hvm,CpuOptions: {CoreCount: 1,ThreadsPerCore: 1},CapacityReservationSpecification: {CapacityReservationPreference: open,},Licenses: [],MetadataOptions: {State: pending,HttpTokens: optional,HttpPutResponseHopLimit: 1,HttpEndpoint: enabled}}
2020-12-31 15:01:39.249+0000 [id=140760]	INFO	h.p.ec2.EC2RetentionStrategy#start: Start requested for EC2 (AWS-connector--eu-west-1(Ireland)((DEV)) - test (i-idididid)
2020-12-31 15:01:39.249+0000 [id=151193]	INFO	hudson.plugins.ec2.EC2Cloud#log: Launching instance: i-idididid`

**Pull request merged**
